petdeface/petdeface.py

```diff
@@ -127,63 +149,67 @@ def deface(args: Union[dict, argparse.Namespace]):
     else:
         args = args
 
-    if os.path.exists(args.bids_dir):
-        # it gets really annoying when there are both gzipped and unzipped nifti files, we are going to
-        # zip them all up before we get started
-        layout = bids.BIDSLayout(args.bids_dir, validate=False)
-        for file in layout.get(extension='nii', return_type='file'):
-            zip_nifti(file)
-
-        if not args.skip_bids_validator:
-            layout = bids.BIDSLayout(args.bids_dir, validate=True)
-        else:
-            layout = bids.BIDSLayout(args.bids_dir, validate=False)
+    if not check_valid_fs_license():
+        raise Exception("You need a valid FreeSurfer license to proceed!")
+
+    participants = args.participant_label or collect_participants(
+        args.bids_dir, bids_validate=~args.skip_bids_validator
+    )
+
+    bidsdata, layout = collect_data(
+        args.bids_dir,
+        participant_label=participants,
+        bids_validate=~args.skip_bids_validator,
+        bids_filters={
+            "fmap": {"datatype": "pet", "suffix": "pet"}
+        },  # this is a hack as the queries coded in collect_data don't include pet
+    )
+    # rename fmap to pet
+    bidsdata["pet"] = bidsdata.pop(
+        "fmap"
+    )  # better solution would be to add this properly in niworkflows collect_data queries
+
+    # Check that PET and T1w files are matched
+    match_error = "Make sure that each PET has a corresponding T1w in the same session."
```

Comment:
I think my latest commit achieves the desired behavior. My testing has been limited, so hopefully it covers all possible situations 🤞.

There are ambiguous cases. Let's say a subject has t1w for sessions 1, 2, and 4. Which t1w should be used for pet in session 3? I didn't add anything in my code to "break ties" and favor one of these sessions over the others. (In current implementation, all of these t1ws are tied for best match, and the code would pick ses-01_t1w to process ses-03_pet just because it comes first in the list.)

Reply:
Awesome - thanks @bilgelm! @bendhouseart we should go ahead and make the test cases to verify that this indeed covers the different cases. Regarding the ambiguous cases, workflows like fmriprep and smriprep go ahead an create a single anatomical image using mri_robust_register if more than one anatomical file is available. Difficult to say what would be the optimal usage here, but I would be fine if there is match within a session between PET and anatomical then use that, and if there is no anatomical present, then use the first available anatomical.